bin/observer.rs

```diff
@@ -0,0 +1,113 @@
+        loop {
+            match sub.next().await {
+                Some(msg) => {
+                    if !is_event_we_care_about(&msg.payload) {
+                        continue;
+                    }
+                    let lattice_id = match self.parser.parse(&msg.subject) {
+                        Some(id) => id,
+                        None => {
+                            trace!(subject = %msg.subject, "Found non-matching lattice subject");
+                            continue;
+                        }
+                    };
+
+                    // Create the reaper for this lattice. This operation returns early if it is
+                    // already running
+                    self.reaper.observe(lattice_id);
+
+                    let needs_command = !self.command_manager.has_consumer(&msg.subject).await;
+                    let needs_event = !self.event_manager.has_consumer(&msg.subject).await;
+                    let client = if needs_command || needs_event {
+                        match self.client_builder.get_connection(lattice_id).await {
+                            Ok(c) => c,
+                            Err(e) => {
+                                error!(error = %e, %lattice_id, "Couldn't construct control client for consumer. Will retry on next heartbeat");
+                                continue;
+                            }
+                        }
+                    } else {
+                        continue;
+                    };
+                    if needs_command {
+                        debug!(%lattice_id, subject = %msg.subject, "Found unmonitored lattice, adding command consumer");
+                        self.command_manager.add_for_lattice(&DEFAULT_COMMANDS_TOPIC.replace('*', lattice_id), lattice_id, CommandWorker::new(client.clone())).await.unwrap_or_else(|e| {
+                            error!(error = %e, %lattice_id, "Couldn't add command consumer. Will retry on next heartbeat");
+                        })
+                    }
+                    if needs_event {
+                        debug!(%lattice_id, subject = %msg.subject, "Found unmonitored lattice, adding event consumer");
+                        self.event_manager.add_for_lattice(&msg.subject, lattice_id, EventWorker::new(self.store.clone(), client)).await.unwrap_or_else(|e| {
+                            error!(error = %e, %lattice_id, "Couldn't add event consumer. Will retry on next heartbeat");
+                        })
+                    }
+                }
+                None => {
+                    warn!("Observer subscriber hang up. Attempting to restart");
+                    sub = get_subscriber(&self.client, subscribe_topics.clone()).await?;
+                }
+            }
+        }
+    }
+}
+
+// This is a stupid hacky function to check that this is a host started or host heartbeat event
+// without actually parsing
+fn is_event_we_care_about(data: &[u8]) -> bool {
```

Comment:
Did you just not want to depend on `cloud_event`? I think we've done `serde_json::from_slice::<CloudEvent>(data).and_then(|evt| evt.ty())` in other places

Reply:
Was trying to avoid the penalty of parsing and allocating just to check type